objstorage/shared.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 The LevelDB-Go and Pebble Authors. All rights reserved. Use
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
+package objstorage
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/pebble/internal/base"
+	"github.com/cockroachdb/pebble/objstorage/sharedobjcat"
+)
+
+// sharedInit initializes the shared object subsystem and finds any shared
+// objects.
+func (p *Provider) sharedInit() error {
+	catalog, contents, err := sharedobjcat.Open(p.st.FS, p.st.FSDirName)
+	if err != nil {
+		return errors.Wrapf(err, "pebble: could not open shared object catalog")
+	}
+	p.shared.catalog = catalog
+	// The creator ID may or may not be initialized yet.
+	p.shared.creatorID = contents.CreatorID
+
+	if p.st.Shared.CreatorID.IsSet() {
+		if err := catalog.SetCreatorID(p.st.Shared.CreatorID); err != nil {
+			catalog.Close()
+			return err
+		}
+		p.shared.creatorID = p.st.Shared.CreatorID
+	}
+
+	for _, meta := range contents.Objects {
+		o := ObjectMetadata{
+			FileNum:  meta.FileNum,
+			FileType: meta.FileType,
+		}
+		o.Shared.CreatorID = meta.CreatorID
+		o.Shared.CreatorFileNum = meta.CreatorFileNum
+		p.mu.knownObjects[o.FileNum] = o
+	}
+	return nil
+}
+
+func (p *Provider) sharedSync() error {
+	batch := func() sharedobjcat.Batch {
+		p.mu.Lock()
+		defer p.mu.Unlock()
+		res := p.mu.shared.catalogBatch.Copy()
+		p.mu.shared.catalogBatch.Reset()
+		return res
+	}()
+
+	if batch.IsEmpty() {
+		return nil
+	}
+
+	err := p.shared.catalog.ApplyBatch(batch)
+	if err != nil {
+		// We have to put back the batch (for the next Sync).
+		p.mu.Lock()
+		defer p.mu.Unlock()
+		batch.Append(p.mu.shared.catalogBatch)
+		p.mu.shared.catalogBatch = batch
+		return err
+	}
+
+	return nil
+}
+
+func (p *Provider) sharedPath(meta ObjectMetadata) string {
+	return "shared://" + sharedObjectName(meta)
+}
+
+func sharedObjectName(meta ObjectMetadata) string {
+	// TODO(radu): prepend a "shard" value for better distribution within the bucket?
```

Comment:
(Huh, wrote a comment here that it ate up.)

If you plan on doing this imminently, it might make sense to do this right now, and maybe leave a TODO to optimize the sharding? We could also have the path go `fmt.Sprintf(%s/%d/%s, creatorID, shard, base.MakeFilename(...))`, that way all the objects for one creatorID are prefixed by that creatorID for easy cleanup, and we use the `/` separator for easy browsing on the s3/gcs web interfaces as those will default to treating those as directories.